x/tokenomics/module/abci.go

```diff
@@ -22,6 +23,17 @@ func EndBlocker(ctx sdk.Context, k keeper.Keeper) error {
 		logger.Error(fmt.Sprintf("could not settle pending claims due to error %v", err))
 		return err
 	}
+
+	defer telemetry.ClaimCounter(
+		telemetry.ClaimProofStageSettled,
+		func() uint64 { return numClaimsSettled },
+	)
+
+	defer telemetry.ClaimCounter(
+		telemetry.ClaimProofStageExpired,
+		func() uint64 { return numClaimsExpired },
+	)
+
 	logger.Info(fmt.Sprintf("settled %d claims and expired %d claims", numClaimsSettled, numClaimsExpired))
 
 	// Update the relay mining difficulty for every service that settled pending
```

Comment:
There's one edge case that's annoying.

If `k.UpdateRelayMiningDifficulty` fails, we'll still have telemetry counting successful claims, while the whole `EndBlocker` is reverting.

This affects both the `ClaimCounter`s above but also whatever was metered in `k.SettlePendingClaims`.

I think these metrics should either introduce a `isSuccess` indicator or be called after `UpdateRelayMiningDifficulty` succeeds.

For whatever was metered in `k.SettlePendingClaims`, I can't come-up with a solution other than accumulating anonymous functions then calling them after `UpdateRelayMiningDifficulty` succeeds. It will be polluting the arguments list, or the return with a function slice though. :thinking:

Reply:
Got some ideas from ChatGPT: https://chatgpt.com/share/a0c91646-0c1b-4cac-8f77-2f8604ab4722

Reply:
You can also use a named return value `err error` and check if its nil or not inside the counter function.